subsys/usb/class/bt_h4.c

```diff
@@ -0,0 +1,325 @@
+	},
+
+	/* Data Endpoint OUT */
+	.if0_out_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_OUT_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(CONFIG_BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+
+	/* Data Endpoint IN */
+	.if0_in_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_IN_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(CONFIG_BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+};
+
+#define H4_CMD 0x01
+#define H4_ACL 0x02
+#define H4_SCO 0x03
+#define H4_EVT 0x04
+
+static struct usb_ep_cfg_data bt_h4_ep_data[] = {
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_OUT_EP_ADDR,
+	},
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_IN_EP_ADDR,
+	},
+};
+
+static void bt_h4_read(u8_t ep, int size, void *priv)
+{
+	struct net_buf *buf = priv;
+
+	if (size > 0) {
+		buf->len += size;
+		net_buf_put(&tx_queue, buf);
+		buf = NULL;
+	}
+
+	if (buf) {
+		net_buf_unref(buf);
+	}
+
+	buf = net_buf_alloc(&tx_pool, K_FOREVER);
+	__ASSERT_NO_MSG(buf);
+
+	net_buf_reserve(buf, BT_BUF_RESERVE);
+
+	/* Start a new read transfer */
+	usb_transfer(bt_h4_ep_data[BT_H4_OUT_EP_IDX].ep_addr, buf->data,
+		     BUF_SIZE, USB_TRANS_READ, bt_h4_read, buf);
+}
+
+static void bt_h4_send(struct net_buf *buf)
+{
+	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_IN:
+		net_buf_push_u8(buf, H4_ACL);
+		break;
+	case BT_BUF_EVT:
+		net_buf_push_u8(buf, H4_EVT);
+		break;
+	default:
+		LOG_ERR("Unknown type %u", bt_buf_get_type(buf));
+		net_buf_unref(buf);
+		return;
+	}
+
+	usb_transfer_sync(bt_h4_ep_data[BT_H4_IN_EP_IDX].ep_addr,
+			  buf->data, buf->len, USB_TRANS_WRITE);
+
+	net_buf_unref(buf);
+}
+
+static void hci_rx_thread(void)
+{
+	LOG_DBG("Start USB Bluetooth thread");
+
+	while (true) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+		bt_h4_send(buf);
+	}
+}
+
+static void bt_h4_recv(struct net_buf *buf)
+{
+	u8_t type;
+
+	type = net_buf_pull_u8(buf);
+
+	switch (type) {
+	case H4_CMD:
+		bt_buf_set_type(buf, BT_BUF_CMD);
+		break;
+	case H4_ACL:
+		bt_buf_set_type(buf, BT_BUF_ACL_OUT);
+		break;
+	default:
+		LOG_ERR("Unknown H4 type %u", type);
+		return;
+	}
+
+	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+	if (bt_send(buf)) {
+		LOG_ERR("Error sending to driver");
+		net_buf_unref(buf);
+	}
+}
+
+static void hci_tx_thread(void)
+{
+	while (true) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&tx_queue, K_FOREVER);
+		bt_h4_recv(buf);
+	}
+}
+
+static void bt_h4_status_cb(struct usb_cfg_data *cfg,
+			    enum usb_dc_status_code status, const u8_t *param)
+{
+	ARG_UNUSED(cfg);
+
+	/* Check the USB status and do needed action if required */
+	switch (status) {
+	case USB_DC_ERROR:
```

Comment:
You can remove unused cases if you like, I do not think that every class have to broadcast it :-)

Reply:
This is just used for logging purpose though but Im fine having a default that just print unhandled state or something like that.